### ISSUE TITLE
Why is the default "stride" of common.DetectMultiBackend set to 64

### ISSUE THREAD

**Issue by MarineCao** [ultralytics/yolov5#7351]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi, I'm trying to run `detect.py`  with my onnx model, while I find that the image size is changed from (416, 416) to (448,448) by the `check_img_size` function, since `model.stride=64`. However, my model is a **P5** model. Then, I find that the default `stride` is set to 64 in `DetectMultiBackend `as follows:
`stride, names = 64, [f'class{i}' for i in range(1000)]  # assign defaults`
If the input is a .pt model, the `stride` will be changed by 
`stride = max(int(model.stride.max()), 32)  # model stride`
while, if the input is a .onnx model, the `stride` will never be changed. Then, if I exported the onnx mdoel by `python export.py --device 0 --img 416`, during inference,  I got this error:
```
onnxruntime.capi.onnxruntime_pybind11_state.InvalidArgument: [ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Got invalid dimensions for input: images for the following indices
 index: 2 Got: 448 Expected: 416
 index: 3 Got: 448 Expected: 416
 Please fix either the inputs or the model.

Process finished with exit code 1
```

Of course, if I use `--dynamic` to export .onnx model, the `detect.py` will run without error, but the image size is still changed to (448,448) and the inference speed is lower. And the the output anchors will be misplaced. 
If I modify the default `stride` to 32, then all the outputs are correct.
So I'm confused about the default setting of stride. Is this a bug？ Or, is there a bug in my code?

Environment:
`Win11 + anaconda + cuda 11.0 + python 3.8 + pytorch 1.7.1 + yolov5-6.0
`

**Comment by glenn-jocher** [comment]
@MarineCao thanks for the question! Some model formats (like PyTorch) can access attached metadata (like stride) to dynamically select the minimum viable stride, while other formats (like ONNX) do not have any metadata and thus lack stride information, in which case stride 64 is assumed since it is valid for both P5 models (minimum stride 32) and P6 models (minimum stride 64).

The only alternative here is to run an image and compare the input to the output to empirically attempt a stride determination at inference time, which would add delay on DetectMultiBackend init. I'll think this over a bit and see what we can do here.

**Comment by glenn-jocher** [comment]
@MarineCao good news 😃! Your original issue may now be fixed ✅ in PR #7353. This PR attaches `stride` and `names` as ONNX model metadata which are then read and used during inference.

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by MarineCao** [comment]
@glenn-jocher Thanks for your quick reply. It works for me！
